src/irc_bridge.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the IRC bridge class to represent the matrix-appservice-app plugin for Synapse."""
+
+# disabling due the fact that collect status does many checks
+# pylint: disable=too-many-return-statements
+
+import logging
+import typing
+
+import ops
+
+import synapse
+from charm_state import CharmState
+
+logger = logging.getLogger(__name__)
+
+IRC_SERVICE_NAME = "irc"
+
+
+class IRCBridge(ops.Object):  # pylint: disable=too-few-public-methods
+    """A class representing the IRC bridge plugin for Synapse application.
+
+    See https://github.com/matrix-org/matrix-appservice-irc/ for more details about it.
+    """
+
+    def __init__(self, charm: ops.CharmBase, charm_state: CharmState):
+        """Initialize a new instance of the IRC bridge class.
+
+        Args:
+            charm: The charm object that the IRC bridge instance belongs to.
+            charm_state: Instance of CharmState.
+        """
+        super().__init__(charm, "irc-bridge")
+        self._charm = charm
+        self._charm_state = charm_state
+        self.framework.observe(charm.on.collect_unit_status, self._on_collect_status)
+
+    @property
+    def _pebble_service(self) -> typing.Any:
+        """Return instance of pebble service.
+
+        Returns:
+            instance of pebble service or none.
+        """
+        return getattr(self._charm, "pebble_service", None)
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
+        """Collect status event handler.
+
+        Args:
+            event: Collect status event.
+        """
+        # pylint:disable=duplicate-code
+        if not self._charm_state.synapse_config.enable_irc_bridge:
+            return
+        container = self._charm.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self._charm.unit.status = ops.MaintenanceStatus("Waiting for Synapse pebble")
```

Comment:
I think it should be event.add_status(ops.MaintenanceStatus("Waiting for Synapse pebble")), otherwise it can be overwritten by the other collect status.

Reply:
I'm not sure tbh - the whole logic in this file is interesting:
Ops will send the collect status at the end of a hook (and we assume synapse should be running after the hook), then we check here whether it is running - if not, we set a global Maintenance. If it is up, to me any other part that checks that synapse is up should be able to clear that Maintenance. Further down we set the status to irc bridge with an `event.add`, after running the enable_irc_bridge (where we again check that synapse is running...). This would need refactoring imo.

Reply:
Not sure neither TBH. We set it here in MaintenanceStatus (but the irc bridge is not configured), and this handler ends here. If any other collect_status handler runs a `add_status(ActiveStatus)`, it will end in active, but the irc bridge is not enabled. What should be the status?

Reply:
Pinging @gregory-schiano as he pointed me to this comment https://github.com/canonical/synapse-operator/pull/30#discussion_r1324745492